timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -747,20 +747,31 @@ template void __global__ k_setup_destination_weights_for_targeted<double>(
 );
 
 void __global__ k_adjust_sample_idxs(
+    const int total_proposals,
     const int batch_size,
+    const int *__restrict__ noise_offset,           // [1]
     const int *__restrict__ targeting_inner_volume, // [batch_size]
     const int *__restrict__ inner_count,            // [1]
-    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const int *__restrict__ partitioned_indices,    // [K]
     int *__restrict__ sample_idxs                   // [batch_size]
 ) {
-    const int local_inner_count = inner_count[0];
+    const int current_offset = *noise_offset;
+    const int local_inner_count = *inner_count;
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
 
     while (idx < batch_size) {
-        const int target_inner = targeting_inner_volume[idx];
-        const int offset = target_inner == 1 ? local_inner_count : 0;
-        const int before = sample_idxs[idx];
-        sample_idxs[idx] = partitioned_indices[before + offset];
+        if (idx + current_offset >= total_proposals) {
+            // The value here is not important, as the result will be ignored in subsuequent kernels
+            // This is done to avoid the fact that the original sampled indices may contain invalid state
+            // thanks to the way the batching performed.
+            sample_idxs[idx] = 0;
```

Comment:
Curiosity: what's the effect of setting the value here, if the value is unimportant / ignored? Naively it seems like just guarding the code that is now in the "else" block would be sufficient.

Reply:
You are right, don't actually need to change it, improved in https://github.com/proteneer/timemachine/pull/1311/commits/bbd65efa03a76cd6094d22c256bfd68b17d1c2cb

Also fixes a bug if the number of blocks was not `ceil_divide(num_samples, tpb)`